docs/architecture/adr-002-light-client-implementation.md

```diff
@@ -0,0 +1,784 @@
+        let (prefix_map, encoded_key) = match path {
+            Path::ClientState(p) => ("client_state_map", p.try_to_vec()),
+            Path::ClientConsensusState(p) => ("consensus_state_map", p.try_to_vec()),
+            Path::Connection(p) => ("connection_end_map", p.try_to_vec()),
+            Path::ChannelEnd(p) => ("channel_end_map", p.try_to_vec()),
+            Path::SeqSend(p) => ("send_sequence_map", p.try_to_vec()),
+            Path::SeqRecv(p) => ("recv_sequence_map", p.try_to_vec()),
+            Path::SeqAck(p) => ("ack_sequence_map", p.try_to_vec()),
+            Path::Commitment(p) => ("packet_commitment_map", p.try_to_vec()),
+            Path::Ack(p) => ("packet_ack_map", p.try_to_vec()),
+            Path::Receipt(p) => ("packet_receipt_map", p.try_to_vec()),
+            Path::UpgradeClient(p) => match p {
+                UpgradeClientPath::UpgradedClientState(_) => {
+                    ("upgraded_client_state_map", p.try_to_vec())
+                }
+                UpgradeClientPath::UpgradedClientConsensusState(_) => {
+                    ("upgraded_consensus_state_map", p.try_to_vec())
+                }
+            },
+            _ => Err(ClientError::Other {
+                description: "unsupported path".into(),
+            })?,
+        };
+
+        let encoded_key = encoded_key.map_err(|_| ClientError::Other {
+            description: "failed to encode key".into(),
+        })?;
+
+        let key_bytes = compute_key_bytes(prefix, prefix_map, encoded_key);
+
+        Ok(jmt::KeyHash::with::<sha2::Sha256>(key_bytes.as_slice()))
+    }
+    ```
+
+### ClientStateValidation methods
+
+- [`verify_client_message()`](https://github.com/informalsystems/sovereign-ibc/blob/09a818d57fed253a500f731eec93c4945df243ad/clients/sov-celestia/src/client_state/validation.rs#L64):
+  Contains the verification steps akin to ICS-07 for the `da_header` field
+  contained in the [`Header`](#header) struct. The brief overview of the primary
+  steps involved in the `SovTmHeader` verification are as follows:
+
+  ```rust
+  /// Verifies the IBC header type for the Sovereign SDK rollups, which consists
+  /// of the DA header and the aggregated proof date validation.
+  pub fn verify_header<V, H>(
+      ctx: &V,
+      client_state: &SovTmClientState,
+      header: &SovTmHeader,
+      client_id: &ClientId,
+      verifier: &impl TmVerifier,
+  ) -> Result<(), ClientError>
+  where
+      V: ExtClientValidationContext,
+      V::ConsensusStateRef: Convertible<SovTmConsensusState, ClientError>,
+      H: MerkleHash + Sha256 + Default,
+  {
+      // Checks the sanity of the fields in the header.
+      header.validate_basic::<H>()?;
+
+      header.validate_da_height_offset(
+          client_state.genesis_da_height(),
+          client_state.latest_height_in_sov(),
+      )?;
+
+      verify_da_header::<V, H>(ctx, client_state, &header.da_header, client_id, verifier)?;
```

Comment:
It seems that non-sequential client updates are not allowed. The header must have the client's latest height as a trusted height.

Can you explain why this is assumed?

Reply:
Since [this logic](https://github.com/informalsystems/sovereign-ibc/blob/43de9b2fa02a35def46b2284b826ce9e4095fa7e/clients/sov-celestia/types/src/client_message/header.rs#L93-L104) is not present in the ADR, let's go ahead with merge. And, tackle the discussion in separate issue.